File: TaylorCode/app/src/main/java/test/taylor/com/taylorcode/ui/custom_view/selector/Selector.java
package test.taylor.com.taylorcode.ui.custom_view.selector;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import test.taylor.com.taylorcode.R;

/**
 * it is a customized view acts like a checkbox.
 * it can be selected or unselected, the background will change accordingly. wrapping this business logic into a single view for clean code in Fragment
 */
public abstract class Selector extends FrameLayout implements View.OnClickListener {
    /**
     * click listener for Selector
     */
    private OnSelectorStateListener stateListener;
    /**
     * the unique tag for a selector
     */
    private String tag;
    /**
     * the group which this Selector belongs to
     */
    private SelectorGroup selectorGroup;

    public Selector(Context context) {
        super(context);
        initView(context, null);
    }

    public Selector(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context, attrs);
    }

    public Selector(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context, attrs);
    }

    private void initView(Context context, AttributeSet attrs) {
        //inflate views
        View view = onCreateView();
        LayoutParams params = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        this.addView(view, params);
        this.setOnClickListener(this);

        //read declared attributes
        if (attrs != null) {
            TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.Selector);
            String text = typedArray.getString(R.styleable.Selector_text);
            int iconResId = typedArray.getResourceId(R.styleable.Selector_img, 0);
            int selectorResId = typedArray.getResourceId(R.styleable.Selector_indicator, 0);
            int textColor = typedArray.getColor(R.styleable.Selector_text_color, Color.parseColor("#FF222222"));
            int textSize = typedArray.getInteger(R.styleable.Selector_text_size, 15);
            tag = typedArray.getString(R.styleable.Selector_tag);
            onBindView(text, iconResId, selectorResId, textColor, textSize);
            typedArray.recycle();
        }
    }

    public Selector setSelectorGroup(SelectorGroup selectorGroup) {
        this.selectorGroup = selectorGroup;
        selectorGroup.addSelector(this);
        return this;
    }

    /**
     * bind the resource to the view which is created from {@link #onCreateView()}
     *
     * @param text           the title of Selector
     * @param iconResId      the image of Selector
     * @param indicatorResId the image show when Selector is in selected state
     * @param textColorResId text color of Selector's text
     * @param textSize       text size of Selector's text
     */
    protected abstract void onBindView(String text, int iconResId, int indicatorResId, int textColorResId, int textSize);

    /**
     * design how the selector looks like
     *
     * @return
     */
    protected abstract View onCreateView();

    public String getTag() {
        return tag;
    }

    public Selector setOnSelectorStateListener(OnSelectorStateListener stateListener) {
        this.stateListener = stateListener;
        return this;
    }

    @Override
    public void onClick(View v) {
        boolean isSelect = switchSelector();
        if (selectorGroup != null) {
            selectorGroup.setSelected(this);
        }
        if (stateListener != null) {
            stateListener.onStateChange(this, isSelect);
        }
    }

    public boolean switchSelector() {
        boolean isSelect = this.isSelected();
        this.setSelected(!isSelect);
        onSwitchSelected(!isSelect);
        return !isSelect;
    }

    protected abstract void onSwitchSelected(boolean isSelect);

    @Override
    public int hashCode() {
        return this.tag.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Selector) {
            return ((Selector) obj).tag.equals(this.tag);
        }
        return false;
    }

    public interface OnSelectorStateListener {
        /**
         * it will be invoked when Selector is selected or unselected
         *
         * @param selector the Selector itself
         * @param isSelect the state of Selector
         */
        void onStateChange(Selector selector, boolean isSelect);
    }
}

File: TaylorCode/app/src/main/java/test/taylor/com/taylorcode/ui/custom_view/selector/kt/Selector.kt
package test.taylor.com.taylorcode.ui.custom_view.selector.kt

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.widget.FrameLayout
import java.io.Closeable
import java.util.*
import kotlin.collections.HashMap

/**
 * a ViewGroup that has customized action when selected or unselected, it could be an substitution for [android.widget.CheckBox] and [android.widget.RadioButton]
 * [contentView] describe how do [Selector] looks like,
 * [onSelectChange] describe what effect will be shown after selection state change,
 * [tags] keeps data bean for this [Selector],
 * [group] describe selection mode for [Selector], and the mode could be extended easily.
 */
class Selector @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    FrameLayout(context, attrs, defStyleAttr) {

    /**
     * the values this [Selector] carry
     */
    private var tags = HashMap<Any?, Closeable?>()

    /**
     * the unique identifier for a [Selector]
     */
    var tag: String = "default tag-${UUID.randomUUID()}"

    /**
     * the identifier for the [SelectorGroup] this [Selector] belongs to
     */
    var groupTag: String = "default group tag"

    /**
     * the [SelectorGroup] this [Selector] belongs to
     */
    var group: SelectorGroup? = null

    /**
     * the layout view for this [Selector]
     */
     var contentView: View? = null
        set(value) {
            field = value
            value?.let {
                addView(it, LayoutParams(MATCH_PARENT, MATCH_PARENT))
                setOnClickListener {
                    group?.onSelectorClick(this)
                }
            }
        }

    /**
     * it will be invoked when the selection state of this [Selector] has changed,
     * override it if you want customized effect of selected or unselected
     */
    var onSelectChange: ((Selector, Boolean) -> Unit)? = null

    /**
     * whether this [Selector] is selected
     */
    var isSelecting: Boolean = false

    init {
        contentView?.let {
            addView(it, LayoutParams(MATCH_PARENT, MATCH_PARENT))
            setOnClickListener {
                group?.onSelectorClick(this)
            }
        }
    }

    operator fun <T : Closeable> set(key: Key<T>, closeable: Closeable) {
        tags[key] = closeable
    }

    operator fun <T : Closeable> get(key: Key<T>): T? = (tags.getOrElse(key, { null })) as T

    /**
     * change the selection state of this [Selector]
     */
    fun setSelect(select: Boolean) {
        group?.onSelectorClick(this)
        showSelectEffect(select)
    }

    fun showSelectEffect(select: Boolean) {
        if (isSelecting != select) {
            onSelectChange?.invoke(this, select)
        }
        isSelecting = select
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        clear()
    }

    /**
     * clear the values attached to this [Selector]
     */
    private fun clear() {
        group?.clear()
        tags.forEach { entry ->
            closeWithException(entry.value)
        }
    }

    private fun closeWithException(closable: Closeable?) {
        try {
            closable?.close()
        } catch (e: Exception) {
        }
    }

    override fun hashCode(): Int {
        return tag.hashCode()
    }

    /**
     * the key for data bean of this [Selector]
     */
    interface Key<E : Closeable>
}

